### PR TITLE
Implement dynamic probes

### PR DESCRIPTION
**Pull request by jkroepke**
This is useful on platforms where a team offers a observability Stack and users could use them, e,g define custom fail reges matches without creating a dedicated module for each use case. In combination with the Probe CRD, it would be an powerful option.

* Fixes https://github.com/prometheus/blackbox_exporter/issues/986

### CODE REVIEW

**Comment by jurgenhaas** [comment]
Very nice, that's exactly what I was looking for. Hope this gets accepted and merged soon.

**Comment by dswarbrick** [comment]
I would recommend drawing attention to possibility that this could be abused by malicious actors. Currently, all aspects of a probe are predefined by a (presumably) responsible sysadmin. As such, blackbox_exporters may be used to probe targets that would not normally be accessible (e.g. behind some network perimiter).

This PR would open up the blackbox_exporter to injecting all manner of arbitrary headers / body in requests to arbitrary targets, and this really accentuates the need to adequately secure the blackbox_exporter from unauthorised use.

**Comment by jurgenhaas** [comment]
> this really accentuates the need to adequately secure the blackbox_exporter from unauthorised use.

Isn't that what should be done in any case, even without this enhancement?

**Comment by dswarbrick** [comment]
> Isn't that what should be done in any case, even without this enhancement?

Arguably, yes. But if the blackbox_exporter has only been configured with a concise set of known-innocent probes, probably a lot of instances are not secured (albeit hopefully not accessible from public networks).

**Comment by jkroepke** [comment]
I can understand the points from @dswarbrick and I guess this feature should be an opt-in by default. Which means, an sys admin has explicit enabled this feature.

Then, this should not introduce new issues by default, and targets behind blackbox_exporter remains secured.

**Comment by SuperQ** [comment]
Yes, we already get enough security reports that the blackbox_exporter is allowing the `target` param to make it a proxy.

**Comment by jkroepke** [comment]
There is always the caveat between unsecured instances of blackbox exporter vs. an well secured environment.

@SuperQ To support both sides, could be a opt-in toggle a compromise?

**Comment by jkroepke** [comment]
@electron0zero Do you have an opinion, here?

**Comment by electron0zero** [comment]
> @electron0zero Do you have an opinion, here?
> 
> Since this is opt-in be default, it should fine. For additional network restrictions, I would recommend NetworkPolicies.

this looks like a valid use-case to me, but since it's a fairly big change I would like to hear what the other maintainers have to say, and ideally have a discussion about the feature, and how we want to implement it :)

cc @mem @roidelapluie

File: main.go [hunk 1]
```diff
@@ -50,6 +50,7 @@ var (
 	configFile     = kingpin.Flag("config.file", "Blackbox exporter configuration file.").Default("blackbox.yml").String()
 	timeoutOffset  = kingpin.Flag("timeout-offset", "Offset to subtract from timeout in seconds.").Default("0.5").Float64()
 	configCheck    = kingpin.Flag("config.check", "If true validate the config file and then exit.").Default().Bool()
+	dynamicProbe   = kingpin.Flag("enable.dynamic-probe", "If true, probes can be dynamically configured via GET parameter using /probe/dynamic endpoint.").Default().Bool()
```
Comment (jkroepke) [inline main.go:53]: I'm also open to increase the security of this feature.

I was thinking about an addition flag like `dynamic-probe.authorized-clients` which defaults to 127.0.0.1. 

If someone really wants to have this feature on a public interface, it requires 2 steps (`enable` + `authorized-clients`)

**Comment by jkroepke** [comment]
@mem @roidelapluie do you have an opinion here?

**Comment by jkroepke** [comment]
@electron0zero it seems like, other maintainers does not have any opinion here.

How we want to proceed here?

**Comment by jc36** [comment]
I have already tested and use these dynamic probes. Thank you very much for the work done!
I also added `query:"timeout"` [here](https://github.com/prometheus/blackbox_exporter/pull/1098/files#diff-fe44f09c4d5977b5f5eaea29170b6a0748819c9d02271746a20d81a5f3efca17R197) as it is necessary for some probes. (units are nanoseconds, so 5s = 5000000000)

**Comment by jkroepke** [comment]
@jc36 the blackbox exporter should respect the scrape_timeout from Prometheus. 

https://github.com/prometheus/blackbox_exporter/blob/c908cba63b5256af83d0553e7fb5049f709aaeac/prober/handler.go#L207

You can test this behavior via curl, by using `curl -H "X-Prometheus-Scrape-Timeout-Seconds: 10"`

Do you think an explicit timeout query is still useful?

**Comment by jkroepke** [comment]
@electron0zero do you have an opionion here how we can proceed here

**Comment by jc36** [comment]
I didn't really understand where I should add the header "X-Prometheus-Scrape-Timeout-Seconds" in the scrape-config job settings so that blackbox-exporter waits for a response from the target for no more than the specified time.

```yaml
  scrape_configs:
        - job_name: blackbox_exporter_tcp
          params:
            prober: ['tcp']
            timeout: ['5000000000'] # timeout to wait response from target
            tcp.preferred_ip_protocol: ['ip4']
          scrape_interval: 30s
          scrape_timeout: 10s  # timeout to wait response from blackbox
          metrics_path: /probe/dynamic
          scheme: http
          static_configs:
            - targets:
                - some-target.com:443
          relabel_configs:
            - source_labels: [__address__]
              target_label: __param_target
            - target_label: __address__
              replacement: extmon-blackbox-exporter:9115
```
In any case, the timeout specified together with other parameters will be clearer.

**Comment by jkroepke** [comment]
@SuperQ @electron0zero @mem @roidelapluie Can I assume that we have a common sense here?

**Comment by GuridMa** [comment]
When will this feature be released approximately?
What version to publish to？

**Review by mem — CHANGES_REQUESTED** [review]
Hi,

Thank you for taking the time to contribute to blackbox_exporter.

I have been looking at this code. It seems mostly OK. The goal of the change is not clear, though. It's hard to give you a good code review without knowing what the goal is.

The bit added to README.md tells the user how to use the feature, but not why they would want to use the feature, e.g. when it's appropriate to use one configuration method over the other. I find the example confusing since that's something that you can do with the regular configuration file. Other than embedding the configuration in the Prometheus configuration, I don't see much advantage with this method. In fact, things that are simple using two configuration files become a little awkward (because params is a sequence of key/value pairs, so while it still looks like you are writing YAML, you are in fact not, at least not what you think you are writing, e.g. `prober: [http]` has to be written like that instead of just `prober: http`; I was a bit confused as to why that worked until I remembered how it's being parsed).

Lacking a description of the goal of this change, my guess is that you have some highly dynamic system and you are obtaining the configuration parameters from some other system. If that's the case, I still don't see why it's not possible to generate the regular configuration file with modules and pass the parameters as usual and then issue a reload when the configuration changes.

**Comment by mem** [comment]
Also, I'm not following how this addresses #1050.

**Comment by mem** [comment]
> Also, I'm not following how this addresses #1050.

I should be more verbose.

That issue is talking about not storing the API key in the configuration file.

While I do see that this change would address that in the sense that the configuration file is not required, you would still need to write this in the Prometheus configuration file, as shown in the example.

That issue is basically asking for a secret to be read from a different location (e.g. vault) and then passed to BBE in some way. This change implements "some way". But if you want to scrape BBE using Prometheus, you would still need to write that information to the configuration file, wouldn't you?

**Comment by jkroepke** [comment]
Goal: As part of the platform team, I want a provide a generic probe service which can be use by any team. Each team can setup probes without having the requirement of preregister probes on the probe service. With the PR, Teams can use the Prometheus Operator Scrape CR to configure any probes which are supported by blackbox exporter.

**Comment by SuperQ** [comment]
> With the PR, Teams can use the Prometheus Operator Scrape CR to configure any probes which are supported by blackbox exporter. Including expecting specific strings, set headers, expect status codes

I agree with @mem, this seems like an XY Problem.

It sounds like what you need is "`ProbeConfig`" CRD that allows teams to dynamically configure the exporter. This way a matching `Probe` would be simpler to maintain.

**Comment by jkroepke** [comment]
The an another goal of the PR is to provide an generic request interface can eliminate the concept additional abstraction of probes. I would like to have one place where I can configure probe targets and probe options. Like the [nagios check `check_http`](https://nagios-plugins.org/doc/man/check_http.html) in the old days. The configuration of probes can be fully shifted to the Prometheus configuration.

> It sounds like what you need is "`ProbeConfig`" CRD

And with this PR, an ProbeConfig CRD is not longer needed, since the configuration could be included into Probe CRs.

The alternative to Prometheus Operator would the [classical annotation based service discovery](https://github.com/prometheus-community/helm-charts/blob/e64c7beef2c09fda29121555ee90f323c25bfebe/charts/prometheus/values.yaml#L988-L1014). Annotation could be use to configure additional probe settings, e.g. expected status code.

**Comment by jkroepke** [comment]
I resolved the merge conflicts.

One more use-case: Since blackbox exporter is embedded into [grafana/agent](https://grafana.com/docs/agent/latest/flow/reference/components/prometheus.exporter.blackbox/), it would be helpful, too.

Instead having the requirements of maintain 2 distinct configs (probes + probe target), the PR aims to avoids this. 

@SuperQ @electron0zero @mem How we can proceed here? Are we comfortable to merge this?

**Comment by SuperQ** [comment]
No, sorry, I don't think we want this feature.

**Comment by electron0zero** [comment]
I am in agreement with @SuperQ and @mem on this :)

**Comment by jkroepke** [comment]
It make me unhappy to read this.

It removes a lot of complexity on a lot of setups by shifting the probe configs into prometheus (or upstream SD).  

Feature like `http.fail_if_body_matches_regexp` are mostly valid for single endpoints and worst case I have to configure a module for each probe to observe (In cases where I have to check more than status code).

@SuperQ I would like to appreciate one time more here to understand to user point of view here. We have the Prometheus Operator Eco system where someone can define probes via Kubernetes Custom Resources, but we do not have a corresponding blackbox_exporter Operator for Kubernetes and I expect in the next 12-24 months, there wont be one. 

Looking at exporters like https://github.com/webdevops/azure-metrics-exporter#default-template , they allows each input as GET parameters, which perfectly integrate into the existing Prometheus Operator eco system.

**Comment by SuperQ** [comment]
> We have the Prometheus Operator Eco system where someone can define probes via Kubernetes Custom Resources, but we do not have a corresponding blackbox_exporter Operator for Kubernetes and I expect in the next 12-24 months, there wont be one.

Maybe you can contribute them.

**Comment by SuperQ** [comment]
> It removes a lot of complexity on a lot of setups by shifting the probe configs into prometheus (or upstream SD).

That's just shifting complexity around. It doesn't actually solve the underlying problem and makes it worse for everyone since there's now a much larger abuse potential.

The XY Problem here is that you need to make the probes self-service. That's a problem with the blackbox_exporter config, not Prometheus.

I recommend checking the Prometheus Operator issue tracker and filing an issue there first.

**Comment by SuperQ** [comment]
Since all of the maintainers agree that we can not accept this, I am going to close this PR.

**Pull request closed without merging**